Q1 Dataset Exploration

Perform the following exploratory steps:
1. Display the first 10 rows
2. Display the dataset structure using info()
3. Find the number of rows and columns
4. Identify the data types of each column


In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# 1. Display the first 10 rows
print("--- FIRST 10 ROWS ---")
print(df.head(10))

# 2. Display the dataset structure using info()
print("\n--- DATASET STRUCTURE ---")
df.info()

# 3. Find the number of rows and columns (Shape)
print("\n--- SHAPE OF THE DATASET ---")
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")

# 4. Identify the data types of each column
print("\n--- COLUMN DATA TYPES ---")
print(df.dtypes)

--- FIRST 10 ROWS ---
   Order_ID  Order_Date Customer       City Region   Product     Category  \
0      1001  2024-10-01    Priya  Hyderabad   East     Mouse  Accessories   
1      1002  2024-05-03    Sneha    Kolkata   East   Monitor  Electronics   
2      1003  2024-01-26    Anita     Mumbai   West     Phone  Electronics   
3      1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories   
4      1005  2024-10-15    Rahul  Bangalore  South   Monitor  Electronics   
5      1006  2024-07-04    Pooja  Bangalore   West   Printer       Office   
6      1007  2024-06-24      Dev    Chennai   East   Monitor  Electronics   
7      1008  2024-03-18    Kavya    Kolkata   East     Phone  Electronics   
8      1009  2024-10-31    Kavya  Ahmedabad   East     Phone  Electronics   
9      1010  2024-12-25    Pooja  Ahmedabad  North     Phone  Electronics   

   Sales  Quantity  Discount  Profit  
0   1735         7      0.00  307.52  
1    807         2      0.15   49.89  
2   2328     

Q2 Sales Filtering Analysis

Filter the dataset to display:
1. Orders where Sales > 3500
2. Orders where Region = "East" and Sales > 2500

**Explain** what insights you observe

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# 1. Orders where Sales > 3500
sales_high = df[df['Sales'] > 3500]
print("--- ORDERS WITH SALES > 3500 (FIRST 5 ROWS) ---")
print(sales_high.head())
print(f"Total rows found: {len(sales_high)}")

# 2. Orders where Region = 'East' and Sales > 2500
east_high_sales = df[(df['Region'] == 'East') & (df['Sales'] > 2500)]
print("\n--- EAST REGION ORDERS WITH SALES > 2500 (FIRST 5 ROWS) ---")
print(east_high_sales.head())
print(f"Total rows found: {len(east_high_sales)}")

--- ORDERS WITH SALES > 3500 (FIRST 5 ROWS) ---
    Order_ID  Order_Date Customer       City Region   Product     Category  \
3       1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories   
4       1005  2024-10-15    Rahul  Bangalore  South   Monitor  Electronics   
7       1008  2024-03-18    Kavya    Kolkata   East     Phone  Electronics   
8       1009  2024-10-31    Kavya  Ahmedabad   East     Phone  Electronics   
10      1011  2024-06-30     Neha      Delhi  South   Printer       Office   

    Sales  Quantity  Discount   Profit  
3    3933         3      0.20   -71.66  
4    3550         2      0.20   146.30  
7    4693         7      0.05   918.49  
8    5071         6      0.10   707.55  
10   5977         6      0.00  1578.32  
Total rows found: 104

--- EAST REGION ORDERS WITH SALES > 2500 (FIRST 5 ROWS) ---
    Order_ID  Order_Date Customer       City Region   Product     Category  \
3       1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories  

1. Category Dominance: Out of the 104 orders generating sales greater
than 3500, high-ticket categories like Electronics (56 orders) and Accessories (39 orders) strongly dominate the results. Products like Phones, Keyboards, and Tablets appear most frequently in high-revenue transactions.
2. East Region Share: Out of the 104 orders with sales greater than 3500 across all locations, 48 orders (nearly $46\%$) come from the East region alone when looking at a benchmark of sales $> 2500$. This indicates that the East region is a primary driver of high-value revenue items (particularly Laptops and Tablets).
3. Profitability Variance: High sales do not always mean positive profits. For instance, Order_ID 1004 has high sales ($3,933$), but generated a negative profit ($-71.66$), likely due to overhead or discount metrics which can be investigated further.

Q3 Customer Purchase Analysis

Identify the top 5 customers based on total sales.

Hint: Use GroupBy with aggregation.

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# 1. Group by Customer and aggregate total Sales
customer_sales = df.groupby('Customer')['Sales'].sum().reset_index()

# 2. Sort customers by Sales in descending order and grab the top 5
top_5_customers = customer_sales.sort_values(by='Sales', ascending=False).head(5)

print("--- TOP 5 CUSTOMERS BY TOTAL SALES ---")
print(top_5_customers.to_string(index=False, formatters={'Sales': '${:,.2f}'.format}))

--- TOP 5 CUSTOMERS BY TOTAL SALES ---
Customer      Sales
   Nitin $59,820.00
     Raj $58,416.00
   Rahul $52,291.00
     Dev $48,500.00
    Isha $46,818.00


Q4 Regional Sales Performance

Calculate the total sales generated by each region.

Then identify:


*   Which region generated the highest revenue?



In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# 1. Calculate the total sales generated by each region
region_sales = df.groupby('Region')['Sales'].sum()

print("--- TOTAL SALES BY REGION ---")
for region, sales in region_sales.items():
    print(f"{region:<10} : ${sales:,.2f}")

# 2. Identify the region with the highest revenue using idxmax()
highest_revenue_region = region_sales.idxmax()
highest_revenue_value = region_sales.max()

print("\n--- REVENUE LEADER ---")
print(f"The region that generated the highest revenue is the {highest_revenue_region} region with ${highest_revenue_value:,.2f}.")

--- TOTAL SALES BY REGION ---
East       : $232,675.00
North      : $158,210.00
South      : $183,108.00
West       : $170,227.00

--- REVENUE LEADER ---
The region that generated the highest revenue is the East region with $232,675.00.


Q5 Product Performance Analysis
1. Calculate the total quantity sold for each product.
2. Which product appears to be the most popular?

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# 1. Calculate the total quantity sold for each product
product_qty = df.groupby('Product')['Quantity'].sum().reset_index()

# 2. Sort the products by quantity in descending order to identify the most popular item
product_qty_sorted = product_qty.sort_values(by='Quantity', ascending=False)

print("--- TOTAL QUANTITY SOLD BY PRODUCT ---")
print(product_qty_sorted.to_string(index=False))

# 3. Identify the most popular product
most_popular_product = product_qty_sorted.iloc[0]['Product']
most_popular_quantity = product_qty_sorted.iloc[0]['Quantity']

print("\n--- POPULARITY SUMMARY ---")
print(f"The most popular product is '{most_popular_product}' with a total of {most_popular_quantity} units sold.")

--- TOTAL QUANTITY SOLD BY PRODUCT ---
   Product  Quantity
   Monitor       156
     Phone       151
    Tablet       150
    Laptop       140
     Mouse       122
Headphones       115
  Keyboard       110
   Printer       108

--- POPULARITY SUMMARY ---
The most popular product is 'Monitor' with a total of 156 units sold.


Q6 Multi-Level Grouping
1. Group the dataset by: Region and Category
2. Calculate the total sales for each combination.
3. What insights can you derive from this analysis?

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# 1. Group by Region and Category, then calculate the total sales
grouped_sales = df.groupby(['Region', 'Category'])['Sales'].sum().reset_index()

print("--- TOTAL SALES BY REGION AND CATEGORY ---")
print(grouped_sales.to_string(index=False, formatters={'Sales': '${:,.2f}'.format}))

--- TOTAL SALES BY REGION AND CATEGORY ---
Region    Category       Sales
  East Accessories  $57,646.00
  East Electronics $160,907.00
  East      Office  $14,122.00
 North Accessories  $62,052.00
 North Electronics  $70,023.00
 North      Office  $26,135.00
 South Accessories  $80,914.00
 South Electronics  $86,690.00
 South      Office  $15,504.00
  West Accessories  $64,811.00
  West Electronics  $84,128.00
  West      Office  $21,288.00


$\text{Derived Insights}$

*   Electronics Dominance in the East: Electronics in the East region is by far the highest-performing segment across the entire dataset, generating a staggering $160,907.00 in sales. This single combination contributes a major portion of the company's total business revenue.

*   Office Supplies Lag Globally: The Office category lags behind significantly in every single region. Its weakest performance is in the East region ($14,122.00) and its strongest is in the North region ($26,135.00), showing that office products have low market penetration or lower average order values.

*   Accessories Regional Patterns: Accessories perform exceptionally well in the South region ($80,914.00), outperforming the North, West, and East branches for that specific category.







Q7 Pivot Table Creation

Create a pivot table showing:

*   Rows → Region
*   Columns → Product
*   Values → Total Sales

Which region sells the most laptops?

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# Create the pivot table using pivot_table() with aggfunc='sum' to aggregate sales
sales_pivot = df.pivot_table(index='Region', columns='Product', values='Sales', aggfunc='sum')

print("--- REGIONAL PRODUCT SALES PIVOT TABLE ---")
print(sales_pivot)

# Identify which region sells the most laptops
highest_laptop_region = sales_pivot['Laptop'].idxmax()
highest_laptop_sales = sales_pivot['Laptop'].max()

print("\n--- LAPTOP SALES OVERVIEW ---")
print(f"The region that sells the most Laptops is the '{highest_laptop_region}' region with a total of ${highest_laptop_sales:,.2f} in sales.")

--- REGIONAL PRODUCT SALES PIVOT TABLE ---
Product  Headphones  Keyboard  Laptop  Monitor  Mouse  Phone  Printer  Tablet
Region                                                                       
East          23899     10845   55568    31299  22902  39667    14122   34373
North         15450     20538    4540    14922  26064  30535    26135   20026
South         24111     23480   23029    19167  33323  18218    15504   26276
West          17159     38499   10826    18178   9153  27850    21288   27274

--- LAPTOP SALES OVERVIEW ---
The region that sells the most Laptops is the 'East' region with a total of $55,568.00 in sales.


Q8 Profitability Analysis

*   Calculate the average profit for each product category.
*   Which category is most profitable?


In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# 1. Calculate the average profit for each product category
category_profit = df.groupby('Category')['Profit'].mean().reset_index()

# 2. Sort to easily find the most profitable category
category_profit_sorted = category_profit.sort_values(by='Profit', ascending=False)

print("--- AVERAGE PROFIT BY PRODUCT CATEGORY ---")
print(category_profit_sorted.to_string(index=False, formatters={'Profit': '${:,.2f}'.format}))

# 3. Identify the most profitable category
most_profitable_category = category_profit_sorted.iloc[0]['Category']
highest_average_profit = category_profit_sorted.iloc[0]['Profit']

print("\n--- PROFITABILITY LEADER ---")
print(f"The most profitable category on average is '{most_profitable_category}' with an average profit of ${highest_average_profit:,.2f} per order.")

--- AVERAGE PROFIT BY PRODUCT CATEGORY ---
   Category  Profit
     Office $471.94
Accessories $410.60
Electronics $396.46

--- PROFITABILITY LEADER ---
The most profitable category on average is 'Office' with an average profit of $471.94 per order.


Q9 Discount Impact Study
Analyze the relationship between discount and profit.

Tasks:
1. Calculate the average profit for each discount level
2. Identify whether higher discounts lead to lower profit

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# 1. Calculate the average profit for each unique discount level
discount_impact = df.groupby('Discount')['Profit'].mean().reset_index()

print("--- AVERAGE PROFIT BY DISCOUNT LEVEL ---")
print(discount_impact.to_string(index=False, formatters={'Discount': '{:.2%}'.format, 'Profit': '${:,.2f}'.format}))

--- AVERAGE PROFIT BY DISCOUNT LEVEL ---
Discount  Profit
   0.00% $751.21
   5.00% $531.37
  10.00% $346.46
  15.00% $237.19
  20.00%  $90.03


$\text{Analytical Conclusion}$

$\text{Do higher discounts lead to lower profit?}$ Yes, absolutely.

There is a clear, inverse linear relationship observed here: as the discount level scales up from $0.00\%$ to $20.00\%$, the average net profit per transaction drops dramatically from $751.21 down to just $90.03. This reveals that aggressive discount policies are severely compressing product margins and cutting into net profitability.

Q10 End-to-End Data Wrangling Task

Perform the following workflow:
1. Filter the dataset where Sales > 3000
2. Group data by Region and Product
3. Calculate total sales
4. Reshape the result into a pivot table

Explain the business insights obtained from this analysis.

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# Step 1: Filter the dataset where Sales > 3000
filtered_df = df[df['Sales'] > 3000]

# Steps 2 & 3: Group data by Region and Product, and calculate total sales
grouped_sales = filtered_df.groupby(['Region', 'Product'])['Sales'].sum().reset_index()

# Step 4: Reshape the result into a pivot table
# This maps out high-value transaction distributions seamlessly
wrangled_pivot = filtered_df.pivot_table(index='Region', columns='Product', values='Sales', aggfunc='sum')

print("--- END-TO-END WRANGLED PIVOT TABLE (SALES > 3000) ---")
print(wrangled_pivot.fillna(0)) # Using fillna(0) ensures clean visualization if any combo has zero records

--- END-TO-END WRANGLED PIVOT TABLE (SALES > 3000) ---
Product  Headphones  Keyboard  Laptop  Monitor  Mouse  Phone  Printer  Tablet
Region                                                                       
East          14469      9642   44096    16420  19000  29500     3696   29229
North          5326     15653    4540     5087  22209  24670    17185    3392
South         10026     19249   17950    14921  26928  18218    14143   19613
West          13898     33380    5922    14354   3511  15653    12648   17185


$\text{Business Insights Obtained}$

$\text{Premium Laptop Hub in the East:}$ Even when filtering down exclusively to large bulk or premium-tier orders (Sales $> 3000$), Laptops in the East region bring in an exceptional total revenue of $44,096.00. This highlights the East region as a critical geographic market for high-ticket corporate enterprise or premium computer purchases.

$\text{Accessories Driving High Ticket Baskets:}$ Interestingly, lower-cost items like Keyboards in the West (\$33,380.00) and Mice in the South (\$26,928.00) show prominent transaction figures in the high-value category. Since individual components are cheap, this reveals a high volume of bulk order behavior (e.g., procurement departments fitting out entire office floors at once) rather than lone consumer shopping.

$\text{Underperforming High-Value Categories:}$ Printers in the East ($3,696.00$) and Tablets in the North ($3,392.00$) are barely breaking past the minimum thresholds. Marketing and B2B sales teams should focus promotional bundles or account targeting strategies to improve wholesale corporate pipelines for these specific regional product segments.